Azure-Samples/azure-search-openai-demo-csharp
# Add .NET MAUI Blazor Hybrid frontend app

## Purpose

Adds a .NET MAUI Blazor Hybrid frontend app that re-uses the same Razor Components used in the Blazor Web frontend app. This helps showcase how a .NET MAUI cross-platform app can share the same Azure OpenAI functionality as a web app, and run natively on Windows, macOS, iOS, and Android devices.

Fixes #253

## Does this introduce a breaking change?

[ ] Yes
[x] No

## Pull Request Type

What kind of change does this Pull Request introduce?

[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:

## How to Test

1. Ensure all existing repo pre-reqs are installed for .NET 8 and Visual Studio, including the .NET MAUI workload
1. Clone repo
1. Follow the existing steps in the repo to publish the app to Azure
1. Get the URL of the app that was published to Azure Container Apps (something like `https://MY_HOSTED_APP.example.azurecontainerapps.io/`)
1. Update the file `app/maui-blazor/MauiProgram.cs` to use that URL to set the `client.BaseAddress` value (there is a 'TODO' note in the file)
1. Set the .NET MAUI Blazor Hybrid app as the startup project
1. Run the app on your device of choice (Windows Desktop, iPhone/iPad, MacBook, Android phone/tablet, etc.) and you should see a UI that exactly matches the Blazor Web app, except that the app is running locally (API calls still go out to Azure as usual)
1. Profit

## Other Information

This PR conceptually has a few areas of change:

1. Almost all of the Razor components in the `app/frontend` Blazor Web app were moved to a new shared `app/SharedWebComponents` Razor Class Library (RCL)
2. The Blazor Web app now references that shared code
3. There is a new .NET MAUI Blazor Hybrid project that is mostly just the default template code. Almost all of the brand new code in this PR is standard template code. This new project references the same shared RCL to reuse the exact same functionality and implementation
4. The few notable changes in the apps are:
   1. The PDF viewer logic was put behind a service interface so that it could use a web-based viewer in the Blazor Web app, and in the future a native PDF viewer in the .NET MAUI app
   2. A UI rendering bug was fixed in `app/SharedWebComponents/Pages/Chat.razor` so that it displays properly on narrow screens (this was a pre-existing issue in the web app, but was very noticable on narrow mobile devices)
   3. Some try/catch statements were added in the speech API calls

cc @BretJohnson @mattleibow @maddymontaquila @lutzroeder
